### PR TITLE
[Improvement][TaskInstance] reduce database queries

### LINKED ISSUE #11523: [Improvement][TaskInstance] reduce database queries
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

reduce database queries

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by longtb976**
## Purpose of the pull request
close #11523

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by SbloodyS** [comment]
@longtb976 Please associate the issue.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 1]
```diff
@@ -1730,8 +1736,14 @@ public Map<String, Object> viewTree(User loginUser, long projectCode, long code,
                 // set treeViewDto instances
                 for (int i = limit - 1; i >= 0; i--) {
                     ProcessInstance processInstance = processInstanceList.get(i);
-                    TaskInstance taskInstance = taskInstanceMapper.queryByInstanceIdAndCode(processInstance.getId(),
-                            Long.parseLong(nodeCode));
+                    TaskInstance taskInstance = null;
```
Comment (stalary) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:1739]:
```suggestion
TaskInstance taskInstance = taskInstances.stream().filter(instance -> instance.getTaskCode() == Long.parseLong(nodeCode) && instance.getProcessInstanceId() == processInstance.getId()).findFirst().orElse(null);
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessInstanceServiceImpl.java [hunk 2]
```diff
@@ -780,24 +780,36 @@ public Map<String, Object> viewGantt(long projectCode, Integer processInstanceId
...
-            if (taskInstance == null) {
-                continue;
+        if (!nodeList.isEmpty()) {
+            List<Long> taskCodes = nodeList.stream().map(Long::parseLong).collect(Collectors.toList());
+            List<TaskInstance> taskInstances = taskInstanceMapper.queryByInstanceIdsAndCodes(
+                    Collections.singletonList(processInstanceId), taskCodes
+            );
+            for (String node : nodeList) {
+                TaskInstance taskInstance = null;
+                for (TaskInstance instance : taskInstances) {
```
Comment (stalary) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessInstanceServiceImpl.java:790]: Same as above.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java (hunk 1 above)
Comment (longtb976) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:1739]: For ends the loop early without going through every element

Comment (stalary) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:1739]: > For ends the loop early without going through every element

ok, you are right~

File: dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.java [hunk 3]
```diff
@@ -53,6 +53,9 @@ TaskInstance queryByInstanceIdAndName(@Param("processInstanceId") int processIns
     TaskInstance queryByInstanceIdAndCode(@Param("processInstanceId") int processInstanceId,
                                           @Param("taskCode") Long taskCode);
 
+    List<TaskInstance> queryByInstanceIdsAndCodes(@Param("processInstanceIds") List<Integer> processInstanceIds,
```
Comment (caishunfeng) [inline dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.java]:
```suggestion
    List<TaskInstance> queryByProcessInstanceIdsAndTaskCodes(@Param("processInstanceIds") List<Integer> processInstanceIds,
```

**Review by caishunfeng — APPROVED** [review]
LGTM

**Pull request merged**